pkg/vault/client.go

```diff
@@ -85,16 +119,138 @@ func (b *BankVaultClient) SetToken(secretPath, token string, log logr.Logger) er
 		return err
 	}
 
-	secretData := secret.Data["data"].(map[string]interface{})
+	secretData, ok := secret.Data["data"].(map[string]interface{})
 
-	if secretData[tokenName] != token {
+	if !ok {
+		secretData = make(map[string]interface{})
+	}
+
+	if !ok || secretData[tokenName] != token {
 
-		log.WithName("vault").Info("secrets don't match, re-applying")
+		b.log.WithName("vault").Info("secrets don't match, re-applying")
 
 		secretData[tokenName] = token
 
+		secret.Data["data"] = secretData
+
 		_, err = b.client.RawClient().Logical().Write(queryPath, secret.Data)
 	}
 
 	return err
 }
+
+// RemoveSecrets will remove all the keys bellow the given paths. It will list
+// all secrets of in the path and delete them according to the deletion policy.
+func (b *BankVaultClient) RemoveSecrets(secrets []VaultSecret) error {
+	for _, secret := range secrets {
+		err := b.removeSecret(secret)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// removeSecret will remove the token according to the DeletetionPolicy
+func (b *BankVaultClient) removeSecret(removeSecret VaultSecret) error {
+
+	secrets, err := b.listSecrets(removeSecret.Path)
+	if err != nil {
+		return err
+	}
+
+	for _, secret := range secrets {
+
+		sPath := path.Join(b.secretEngine, "metadata", removeSecret.Path, secret)
+
+		s, err := b.client.RawClient().Logical().Read(sPath)
+		if err != nil {
+			return err
+		}
+
+		versions, err := b.getVersionList(s.Data)
+		if err != nil {
+			return err
+		}
+
+		switch b.deletionPolicy {
+		case synv1alpha1.ArchivePolicy:
+			b.log.Info("soft deleting secret", "secret", removeSecret.Path)
+			err := b.deleteToken(path.Join(b.secretEngine, "delete", removeSecret.Path, secret), versions)
+			if err != nil {
+				return err
+			}
+		case synv1alpha1.DeletePolicy:
+			b.log.Info("destroying secret", "secret", removeSecret.Path)
+			err := b.destroyToken(path.Join(b.secretEngine, "metadata", removeSecret.Path, secret), versions)
+			if err != nil {
+				return err
+			}
+		default:
+			continue
```

Comment:
Should we return an error here?
Silently ignoring seems dangerous.

Reply:
How should we handle the error? Simply logging it? If we just return it until the reconcile, we'll generate an infinity reconcile loop...

Reply:
Yes, return it. The exponential backoff will make sure it's retried with less frequency.